### PR TITLE
Start of out of band work

### PR DESCRIPTION
**Pull request by timbru**
Parses a <publisher_request> XML message and validates the included TA certificate.

So, added basic support for handling XML and Identity Certificates that will be useful for the next things that are needed for supporting the publication (and later provisioning) protocol. This stuff is in rpki-rs because both the publication server and the CA implementations will have to use this.

### CODE REVIEW

File: src/oob/idcert.rs [hunk 1]
```diff
@@ -0,0 +1,526 @@
...
+impl AsRef<IdCert> for IdCert {
+    fn as_ref(&self) -> &Self {
+        self
+    }
+}
+
+//------------ Validity ------------------------------------------------------
+
+#[derive(Clone, Debug)]
+pub struct Validity {
```
Comment (partim) [inline src/oob/idcert.rs]: Perhaps move to ::x509 and replace occurrence in ::cert::Cert?

Also for SubjectPublicKeyInfo below …

File: src/oob/idcert.rs [hunk 2]
```diff
@@ -0,0 +1,526 @@
...
+            self.subject_public_key.octet_slice().unwrap()
+        )
+    }
+}
+
+
+//------------ PublicKeyAlgorithm --------------------------------------------
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum PublicKeyAlgorithm {
```
Comment (partim) [inline src/oob/idcert.rs]: Side note: I want to move all these algorithm related things into a module of their own and let them provide the actual crypto primitives, too, so we can more easily add algorithms later.

File: src/oob/idcert.rs [hunk 3]
```diff
@@ -0,0 +1,526 @@
...
+    use std::fs;
+    use bytes::Bytes;
+    use time;
+    use chrono::{TimeZone, Utc};
+
+    #[test]
+    fn test_parse_id_publisher_ta_cert() {
+        let d = Utc.ymd(2012, 1, 1).and_hms(0, 0, 0);
+        time::with_now(d, || {
+        let data = fs::read("test/oob/id-publisher-ta.cer").unwrap();
```
Comment (partim) [inline src/oob/idcert.rs]: Use the `input!` macro instead of doing an actual file system read?

File: src/oob/xml.rs [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+use xml::reader::XmlEvent;
+use xml::attribute::OwnedAttribute;
+
+
+//------------ XmlReader -----------------------------------------------------
+
+/// A convenience wrapper for RPKI XML parsing
+///
+/// This type only exposes things we need for the RPKI XML structures.
+pub struct XmlReader<R: io::Read> {
```
Comment (partim) [inline src/oob/xml.rs]: I’d suggest to rearrange XmlReader in such a way that its access is similar to the way the ber crate allows access, i.e., via nested calls to a `take_element` method that takes a closure and passes on the XmlReader, ensuring correct nesting of XML elements and such.

File: src/x509.rs [hunk 5]
```diff
@@ -212,6 +213,7 @@ impl Time {
 
     pub fn validate_not_before(&self) -> Result<(), ValidationError> {
         if time::now() < self.0 {
+            println!("Not valid before!");
```
Comment (partim) [inline src/x509.rs]: Probably forgot to take this out?

File: src/x509.rs [hunk 6]
```diff
@@ -268,6 +270,12 @@ fn read_four_char<S: Source>(source: &mut S) -> Result<u32, S::Err> {
 #[derive(Clone, Copy, Debug)]
 pub struct ValidationError;
 
+impl fmt::Display for ValidationError{
```
Comment (partim) [inline src/x509.rs]: `#[derive(Fail)]` and `[fail(display="validation error")]` instead?

File: src/x509.rs (hunk 5 above)
Comment (timbru) [reply src/x509.rs]: now gone

File: src/oob/xml.rs (hunk 4 above)
Comment (timbru) [reply src/oob/xml.rs]: Good suggestion, thanks for the pointers. Done.

File: src/x509.rs (hunk 6 above)
Comment (timbru) [reply src/x509.rs]: done

File: src/oob/idcert.rs (hunk 3 above)
Comment (timbru) [reply src/oob/idcert.rs]: done (also in exchange.rs)

File: src/oob/idcert.rs (hunk 1 above)
Comment (timbru) [reply src/oob/idcert.rs]: Now importing from ::cert::Cert instead.. not sure yet whether this should be moved to ::x509.

**Pull request merged**